odl/solvers/functional/default_functionals.py

```diff
@@ -1326,6 +1376,185 @@ def convex_conj(self):
+
+        # Rotate the axes so the svd-direction is first
+        s_reordered = self._moveaxis(svd_diag, -1, 0)
+
+        # Return nuclear norm
+        return self.outernorm(self.pwisenorm(s_reordered))
+
+    @property
+    def proximal(self):
+        """Return the proximal operator.
+
+        Raises
+        ------
+        NotImplementedError
+            if ``outer_exp`` is not 1 or ``singular_vector_exp`` is not 1, 2 or
+            infinity
+        """
+
+        if self.outernorm.exponent != 1:
+            raise NotImplementedError('`proximal` only implemented for '
+                                      '`outer_exp==1`')
+        if self.pwisenorm.exponent not in [1, 2, np.inf]:
+            raise NotImplementedError('`proximal` only implemented for '
+                                      '`singular_vector_exp` in [1, 2, inf]')
+
+        def nddot(a, b):
+            """Compute pointwise matrix product in the last indices."""
+            return np.einsum('...ij,...jk->...ik', a, b)
+
+        func = self
+
+        class NuclearNormProximalOperator(Operator):
+            """Proximal operator of `NuclearNorm`."""
+            def __init__(self, sigma):
+                self.sigma = sigma
+                Operator.__init__(self, func.domain, func.domain)
+
+            def _call(self, x):
+                """Return ``self(x)``."""
+                arr = func._asarray(x)
+
+                # Compute SVD
+                U, s, Vt = np.linalg.svd(arr, full_matrices=False)
+
+                # transpose pointwise
+                V = Vt.swapaxes(-1, -2)
+
+                # Take pseudoinverse of s
+                sinv = s.copy()
+                sinv[sinv != 0] = 1 / sinv[sinv != 0]
+
+                # Take pointwise proximal operator of s w.r.t. the norm
+                # on the singular vectors
+                if func.pwisenorm.exponent == 1:
+                    sprox = np.sign(s) * np.maximum(np.abs(s) - self.sigma, 0)
+                elif func.pwisenorm.exponent == 2:
+                    s_reordered = func._moveaxis(s, -1, 0)
+                    snorm = func.pwisenorm(s_reordered).asarray()
+                    snorm = np.maximum(self.sigma, snorm)
+                    sprox = (1 - self.sigma / snorm)[..., None] * s
+                elif func.pwisenorm.exponent == np.inf:
+                    snorm = np.sum(np.abs(s), axis=-1)
+                    snorm = np.maximum(self.sigma, snorm)
+                    sprox = (1 - self.sigma / snorm)[..., None] * s
+                else:
+                    raise RuntimeError
+
+                # Compute s matrix
+                sproxsinv = (sprox * sinv)[..., :, None]
+
+                # Compute the final result
+                result = nddot(nddot(arr, V), sproxsinv * Vt)
+
+                # Cast to vector and return. Note array and vector have
+                # different shapes.
+                return func._asvector(result)
```

Comment:
This is really some complex code. I feel like I could really use a mathematical description to understand it better.